File: library/src/main/java/cn/wandersnail/widget/listview/PullRefreshLayout.java
package cn.wandersnail.widget.listview;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.AdapterView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

/**
 * 下拉刷新，控制在顶部才拦截触摸事件
 * <p>
 * date: 2019/8/22 19:12
 * author: zengfansheng
 */
public class PullRefreshLayout extends SwipeRefreshLayout {
    private AdapterView<?> lv;

    public PullRefreshLayout(@NonNull Context context) {
        super(context);
    }

    public PullRefreshLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }
    
    private boolean isListViewNotAtTop() {
        if (lv != null) {
            View view = lv.getChildAt(0);
            return view != null && (lv.getFirstVisiblePosition() != 0 || view.getTop() != 0);
        }
        return true;
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        if (getChildCount() > 0) {
            View view = getChildAt(0);
            if (view instanceof AdapterView<?>) {
                lv = (AdapterView<?>) view;
            }
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        return isListViewNotAtTop() && super.onTouchEvent(ev);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        return isListViewNotAtTop() && super.onInterceptTouchEvent(ev);
    }
}

File: library/src/main/java/com/snail/widget/PullRefreshLayout.kt
package com.snail.widget

import android.content.Context
import android.support.v4.widget.SwipeRefreshLayout
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.AdapterView

/**
 * 描述: 下拉刷新，控制在顶部才拦截触摸事件
 * 时间: 2018/6/8 09:20
 * 作者: zengfansheng
 */
class PullRefreshLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : SwipeRefreshLayout(context, attrs) {
    private var lv: AdapterView<*>? = null

    private val isListViewNotAtTop: Boolean
        get() {
            if (lv != null) {
                val child = lv!!.getChildAt(0)
                return child != null && (lv!!.firstVisiblePosition != 0 || child.top != 0)
            }
            return true
        }

    override fun onFinishInflate() {
        super.onFinishInflate()
        if (childCount > 0) {
            val view = getChildAt(0)
            if (view is AdapterView<*>) {
                lv = view
            }
        }
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {
        return isListViewNotAtTop && super.onTouchEvent(ev)
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        return isListViewNotAtTop && super.onInterceptTouchEvent(ev)
    }
}
